src/makelogline.c

```diff
@@ -53,32 +65,36 @@ void fillto(int n);
  *
  *   See function definitions below
  */
-void makelogline(void) {
+char *makelogline(struct qso_t *qso) {
     char freq_buff[10];
 
+    char *logline = g_malloc0(sizeof(logline4));
+
     /* first fixed (contest independent) part of logline */
-    prepare_fixed_part();
-    assert(strlen(logline4) == 54);
+    prepare_fixed_part(logline, qso);
+    assert(strlen(logline) == 54);
 
     /* second (contest dependent) part of logline */
-    prepare_specific_part();
-    assert(strlen(logline4) == 77);
+    prepare_specific_part(logline, qso);
+    assert(strlen(logline) == 77);
 
     /* add points to logline if in contest */
     if (iscontest && !CONTEST_IS(DXPED)) {
-	sprintf(logline4 + 76, "%2d", qso_points);
+	sprintf(logline + 76, "%2d", qso_points);   //FIXME qso->score
     }
 
-    fillto(80);
+    FILL_TO(80);
 
     /* add freq to end of logline */
-    if (trx_control) {
-	snprintf(freq_buff, 8, "%7.1f", freq / 1000.0);
-	strcat(logline4, freq_buff);
+    if (qso->freq > 0) {
```

Comment:
`trx_control` tells whether rig control is currently in use, whereas `qso->freq > 0` checks if freq was known at the time of the QSO. Generally this could have been earlier or even currently but on another Tlf node. So relying on `trx_control` is not correct.

If trx control is disabled while Tlf is running then `freq` is not set to zero (previous value is kept). Similarly if trx control is enabled then rig connection is not initialized, freq stays zero. IMO both are bugs.